src/evo/dmnstate.h

```diff
@@ -203,27 +218,42 @@ class CDeterministicMNStateDiff
     CDeterministicMNStateDiff() = default;
     CDeterministicMNStateDiff(const CDeterministicMNState& a, const CDeterministicMNState& b)
     {
-#define DMN_STATE_DIFF_LINE(f) if (a.f != b.f) { state.f = b.f; fields |= Field_##f; }
-        DMN_STATE_DIFF_ALL_FIELDS
-#undef DMN_STATE_DIFF_LINE
-        if (fields & Field_pubKeyOperator) { state.nVersion = b.nVersion; fields |= Field_nVersion; }
+        boost::hana::for_each(members, [&](auto&& member) {
+            if (member.get(a) != member.get(b)) {
+                member.get(state) = member.get(b);
+                fields |= member.mask;
+            }
+        });
+        if (fields & Field_pubKeyOperator) {
+            // pubKeyOperator needs nVersion
+            state.nVersion = b.nVersion;
+            fields |= Field_nVersion;
+        }
     }
 
     [[nodiscard]] UniValue ToJson(MnType nType) const;
 
     SERIALIZE_METHODS(CDeterministicMNStateDiff, obj)
     {
+        READWRITE(VARINT(obj.fields));
+
         // NOTE: reading pubKeyOperator requires nVersion
         bool read_pubkey{false};
-        READWRITE(VARINT(obj.fields));
-#define DMN_STATE_DIFF_LINE(f) \
-        if (strcmp(#f, "pubKeyOperator") == 0 && (obj.fields & Field_pubKeyOperator)) {\
-            SER_READ(obj, read_pubkey = true); \
-            READWRITE(CBLSLazyPublicKeyVersionWrapper(const_cast<CBLSLazyPublicKey&>(obj.state.pubKeyOperator), obj.state.nVersion == CProRegTx::LEGACY_BLS_VERSION)); \
-        } else if (obj.fields & Field_##f) READWRITE(obj.state.f);
-
-        DMN_STATE_DIFF_ALL_FIELDS
-#undef DMN_STATE_DIFF_LINE
+        boost::hana::for_each(members, [&](auto&& member) {
+            using BaseType = std::decay_t<decltype(member)>;
+            using T1 = std::decay_t<decltype(member.get(obj.state))>;
+            if constexpr (BaseType::mask == Field_pubKeyOperator) {
+                if (obj.fields & member.mask) {
+                    SER_READ(obj, read_pubkey = true);
+                    READWRITE(CBLSLazyPublicKeyVersionWrapper(const_cast<CBLSLazyPublicKey&>(obj.state.pubKeyOperator), obj.state.nVersion == CProRegTx::LEGACY_BLS_VERSION));
+                }
+            } else {
+                if (obj.fields & member.mask) {
```

Comment:
It looks like conditions in implementation are changed after refactoring:

A = is pubKeyOperator
B = obj.fields & pubkeyOperator
C = obj.fields & mask

`if ((!A || !B) && C)` is changed to `if (!A && C)`

It seems as this change is valid. Let's simplify code furthermore:

```
if (obj.fields & member.mask) {
    if constexpr (BaseType::mask == Field_pubKeyOperator) {
        SER_READ(obj, read_pubkey = true);
        READWRITE(CBLSLazyPublicKeyVersionWrapper(const_cast<CBLSLazyPublicKey&>(obj.state.pubKeyOperator), obj.state.nVersion == CProRegTx::LEGACY_BLS_VERSION));
    } else {
        READWRITE(member.get(obj.state));
    }
}

Reply:
Would this simplification retain the compile-time nature of the `constexpr` evaluation? From what I understand, attempting a `constexpr` evaluation within a non-`constexpr` evaluation will downgrade the compile-time evaluation to runtime.